build: add remark plugin for validating HTML section structure

Planeshifter opened the pull request:
> What is the purpose of this pull request?

This pull request:

-   adds a remark plugin for validating that the HTML sections structure in `README.md` files is accurate.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses https://github.com/stdlib-js/metr-issue-tracker/issues/49 by resolving 1) and 2).

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: add remark plugin for validating HTML section structure

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter commented:
/stdlib make-commands

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,300 @@
+* @returns {void}
+*/
+function linter( tree, file, options, clbk ) {
+	var isValidWhitespace;
```
Why are all these variables declared in this scope and not in the function scope in which they are used?

Planeshifter replied:
Addressed.

Planeshifter committed:
chore: address PR feedback and move repeated logic to helper functions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+* @param {Object} node - AST node containing comment
+* @param {RegExp} matchRegex - regular expression for matching comment
+* @param {RegExp} simpleMatchRegex - alternative regular expression for matching comment
+* @returns {string|null} extracted class name or null if no match
```
```suggestion
* @returns {(string|null)} extracted class name or null if no match
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+	var simpleMatch;
+	var match;
+
+	match = matchRegex.exec( node.value );
```
Do we need to worry about resetting the match index since we are using `exec` here and below?

Planeshifter replied:
No, we don't have global or sticky flag on the regular expressions that are executed here.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+	value = node.value;
+
+	// Look for at least two consecutive newlines (representing an empty line):
+	return (/(\r?\n)\s*(\r?\n)/).test( value );
```
Is there a reason not to move this regexp to the parent scope to be with the other regexp?

Planeshifter replied:
Moved. Thanks for noticing!

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+			return;
+		}
+
+		// Check if this is a section start tag
```
```suggestion
		// Check if this is a section start tag:
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+		}
+
+		// Check if this is a section end tag:
+		endMatch = SECTION_END.exec( node.value );
```
Since we are using `exec`, do we need to worry about resetting the regexp state?

Planeshifter replied:
 No. [Only stateful ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#description) when global or sticky flags are set.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+
+			// Special handling for direct closing tag followed by a comment (no text node between)...
+			if ( index + 1 < parent.children.length ) {
+				nextNode = parent.children[index + 1];
```
```suggestion
				nextNode = parent.children[ index + 1 ];
```

Consistent spacing.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+					commentClass = extractCommentClass( nextNode, END_COMMENT, END_COMMENT_SIMPLE );
+
+					if ( commentClass ) {
+						// Check if there are at least 2 line breaks between them
```
```suggestion
						// Check if there are at least 2 line breaks between them:
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+
+			// Standard handling for other cases with whitespace between section end and comment...
+			if ( index + 2 < parent.children.length ) {
+				nextNode = parent.children[index + 1];
+				commentNode = parent.children[index + 2];
```
```suggestion
				nextNode = parent.children[ index + 1 ];
				commentNode = parent.children[ index + 2 ];
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+					} else {
+						// The HTML node exists but it's not a proper closing comment...
+						debug( 'Invalid closing comment format' );
+						msg = 'Invalid section closing comment format. Expected <!-- /.' + sectionStack[sectionStack.length - 1].className + ' -->';
```
```suggestion
						msg = 'Invalid section closing comment format. Expected <!-- /.' + sectionStack[ sectionStack.length - 1 ].className + ' -->';
```

Consistency with L262.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+				} else {
+					// The node after whitespace is not an HTML node (comment)...
+					debug( 'Node after whitespace is not an HTML comment' );
+					msg = 'Missing section closing comment. Section closing tag should be followed by <!-- /.' + sectionStack[sectionStack.length - 1].className + ' --> comment.';
```
```suggestion
					msg = 'Missing section closing comment. Section closing tag should be followed by <!-- /.' + sectionStack[ sectionStack.length - 1 ].className + ' --> comment.';
```

L262

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,366 @@
+			} else {
+				// There's not enough nodes after the section closing tag...
+				debug( 'Not enough nodes after section end' );
+				msg = 'Missing section closing comment. Section closing tag should be followed by <!-- /.' + sectionStack[sectionStack.length - 1].className + ' --> comment on a separate line.';
```
```suggestion
				msg = 'Missing section closing comment. Section closing tag should be followed by <!-- /.' + sectionStack[ sectionStack.length - 1 ].className + ' --> comment on a separate line.';
```

L262

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (commented):
Other than questions around the use of `exec`, this is shaping up.

Planeshifter committed:
chore: rename variables and define empty line regexp

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js:
```diff
@@ -0,0 +1,367 @@
+						if ( !hasValidWhitespace( node, nextNode ) ) {
+							msg = 'Missing proper empty line after closing section tag. There should be an empty line between </section> and the closing comment.';
+							errCode = 'missing-empty-line-after-section';
+							reportErr( file, node, msg, errCode );
```
@Planeshifter Just checking, but for a number of these `reportErr` statements, we don't early return, but instead fall through to subsequent logic. Are these fall-throughs intentional? I'd want to avoid a scenario where lint errors pile up and it becomes confusing for devs to understand what the actual issue is which is triggering lint failures.

kgryte replied:
If we shouldn't be early returning, it may be good to add a comment in each circumstance indicating that fall-through is intentional.

Planeshifter replied:
@kgryte Double-checked, left some comments, and added an early return to make the code more readable.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into philipp/add-remark-plugin-section-structure

Planeshifter committed:
chore: disable lint directive and shorten variable name

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: add comments

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
refactor: add early return to make code more readable

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM
